funsor/affine.py

```diff
@@ -1,16 +1,103 @@
 from collections import OrderedDict
+from functools import reduce
 
 import opt_einsum
 import torch
+from multipledispatch import dispatch
```

Comment:
nit: use `singledispatch` here instead?  I think it's faster

Reply:
done.